magpie/services.py

```diff
@@ -287,28 +309,63 @@ def _get_acl_cached(self, service_name, request_method, request_path, user_id):
         permissions = self.permission_requested()
         if permissions is None:
             return [DENY_ALL]
-        resource = self.resource_requested()
-        if not resource:
+
+        target_resources = self.resource_requested()
+        if not target_resources:
             return [DENY_ALL]
-        if not isinstance(resource, tuple):
-            is_target = False
-        else:
-            resource, is_target = resource
+        if not isinstance(target_resources, list):
+            if not isinstance(target_resources, tuple):
+                resource = target_resources
+                is_target = False
+            else:
+                resource, is_target = target_resources
+            target_resources = [(resource, is_target)]
+        if any(res[0] is None for res in target_resources):
+            return [DENY_ALL]
+
         if not isinstance(permissions, (list, set, tuple)):
             permissions = {permissions}
         user = self.user_requested()
-        return self._get_acl(user, resource, permissions, allow_match=is_target)
 
-    def _get_acl(self, user, resource, permissions, allow_match=True):
-        # type: (models.User, ServiceOrResourceType, Collection[Permission], bool) -> AccessControlListType
+        return self._get_acl(user, target_resources, permissions)
+
+    def _get_acl(self, user, resources, permissions):
+        # type: (models.User, MultiResourceRequested, Collection[Permission]) -> AccessControlListType
         """
-        Resolves the resource-tree and the user/group inherited permissions into a simplified ACL for this resource.
+        Resolves resource-tree and user/group inherited permissions into simplified :term:`ACL` of requested resources.
+
+        Contrary to :meth:`effective_permissions` that can be resolved only for individual :term:`Resource`, :term:`ACL`
+        is involved during actual request access, which could refer to multiple :term:`Resource` references or distinct
+        :term:`Permission` names if the identified parent :term:`Service` supports it.
+
+        When more than one item is specified for validation (any combination of :term:`Resource` or :term:`Permission`),
+        **ALL** of them must be granted access to resolve as :data:`Access.ALLOW`. Any denied access blocks the whole
```

Comment:
So you're allowing to deny a wfs or a wms and this will result in denying access to both?
This is kind of adding a whole new dimension to this inheritance|allow|deny|group membership puzzle 😜. How this decision is reflected in the UI? In the API reason?

Reply:
No, because only one between WFS/WMS can be targeted. The service first resolves `geoserver/<OWS>`, and then the permisions applicable for the identified `<OWS>`. Also, they would most probably not have the same permission derived from `request=GetMap|GetFeature|...` parameter.

This **ALL(Allow)** concept is more about `layers=layer1,layer2`. Both must be allowed to allow the full request.

Reply:
Sorry, I should have resolved this thread once I added the following comment.